### PR TITLE
fix(npm): support semver with build metadata

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

`semver.clean` removes build metadata which makes us unable to resolve ranges that contain them

Fixes https://github.com/yarnpkg/berry/issues/1445
Fixes https://github.com/yarnpkg/berry/discussions/3980

**How did you fix it?**

Implement our own `clean` method.

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-core/sources/semverUtils.ts [hunk 1]
```diff
@@ -82,3 +82,20 @@ export function validRange(potentialRange: string): semver.Range | null {
...
   return range;
 }
+
+/**
+ The RegExp from https://semver.org/ but modified to
+ - allow the version to start with `(?:[\Wv=]*?)`
+ - allow the version to end with `(?:\W*)`
+ - place the valid version in capture group one
+ */
+const CLEAN_SEMVER_REGEXP = /^(?:[\Wv=]*?)((0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(?:\W*)$/;
```
Comment (paul-soporan) [inline packages/yarnpkg-core/sources/semverUtils.ts:92]: What's the reason for allowing versions such as `1.2.3!` (and cleaning them into `1.2.3`)? :thinking:

Comment (merceyz) [reply packages/yarnpkg-core/sources/semverUtils.ts]: Wow, don't know how/why I mixed up the tokens but it was supposed to be `\s`.

File: packages/yarnpkg-core/sources/semverUtils.ts [hunk 2]
```diff
@@ -82,3 +82,20 @@ export function validRange(potentialRange: string): semver.Range | null {
...
+ - allow the version to end with `(?:\s*)`
+ - place the valid version in capture group one
+ */
+const CLEAN_SEMVER_REGEXP = /^(?:[\sv=]*?)((0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(?:\s*)$/;
+
+/**
+ * Cleans the potential version by removing leading/trailing whitespace and '=v' prefix
+ * @returns A valid SemVer string, otherwise null
+ */
+export function clean(potentialVersion: string): string | null {
```
Comment (arcanis) [inline packages/yarnpkg-core/sources/semverUtils.ts:98]: Can you add a small set of unit tests?

Comment (merceyz) [reply packages/yarnpkg-core/sources/semverUtils.ts:98]: Sure

**Pull request merged**